Source/ASTextNode2.mm

```diff
@@ -783,12 +785,108 @@ - (void)_setHighlightRange:(NSRange)highlightRange forAttributeName:(NSString *)
   // Set these so that link tapping works.
   _highlightedLinkAttributeName = highlightedAttributeName;
   _highlightedLinkAttributeValue = highlightedAttributeValue;
-  _highlightRange = highlightRange;
 
-  AS_TEXT_ALERT_UNIMPLEMENTED_FEATURE();
-  // Much of the code from original ASTextNode is probably usable here.
+  if (!NSEqualRanges(highlightRange, _highlightRange) && ((0 != highlightRange.length) || (0 != _highlightRange.length))) {
 
-  return;
+    _highlightRange = highlightRange;
+
+    if (_activeHighlightLayer) {
+      if (animated) {
+        __weak CALayer *weakHighlightLayer = _activeHighlightLayer;
+        _activeHighlightLayer = nil;
+
+        weakHighlightLayer.opacity = 0.0;
+
+        CFTimeInterval beginTime = CACurrentMediaTime();
+        CABasicAnimation *possibleFadeIn = (CABasicAnimation *)[weakHighlightLayer animationForKey:@"opacity"];
+        if (possibleFadeIn) {
+          // Calculate when we should begin fading out based on the end of the fade in animation,
+          // Also check to make sure that the new begin time hasn't already passed
+          CGFloat newBeginTime = (possibleFadeIn.beginTime + possibleFadeIn.duration);
+          if (newBeginTime > beginTime) {
+            beginTime = newBeginTime;
+          }
+        }
+
+        CABasicAnimation *fadeOut = [CABasicAnimation animationWithKeyPath:@"opacity"];
+        fadeOut.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
+        fadeOut.fromValue = possibleFadeIn.toValue ? : @(((CALayer *)weakHighlightLayer.presentationLayer).opacity);
+        fadeOut.toValue = @0.0;
+        fadeOut.fillMode = kCAFillModeBoth;
+        fadeOut.duration = ASTextNodeHighlightFadeOutDuration;
+        fadeOut.beginTime = beginTime;
+
+        dispatch_block_t prev = [CATransaction completionBlock];
+        [CATransaction setCompletionBlock:^{
+          [weakHighlightLayer removeFromSuperlayer];
+        }];
+
+        [weakHighlightLayer addAnimation:fadeOut forKey:fadeOut.keyPath];
+
+        [CATransaction setCompletionBlock:prev];
```

Comment:
Should we instead push a transaction and just set the completion block of that rather than modifying our parent transaction?

Reply:
Imho let's use the same implementation for `ASTextNode2` we have for `ASTextNode` and improve that on both sides if needed on a follow up PR.